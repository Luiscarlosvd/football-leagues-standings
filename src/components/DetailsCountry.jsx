import { useParams } from "react-router";
import { useEffect, useState } from "react";
import { getCountryDetails } from "../redux/countries/countriesSlice";
import { useSelector, useDispatch } from "react-redux";

const DetailsCountry = () => {
  const country = useSelector((state) => state.countries);
  const { countryName } = useParams();
  const [ currencies, setCurrencies ] = useState([]);
  const filteredCountry = country.countries.find((country) => country.name === countryName);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getCountryDetails(filteredCountry.officialName));
    console.log(country.countryDetails)
  }, [dispatch, countryName]);

  useEffect(() => {
    if (country.detailsStatus === 'fulfilled') {
      if(country.countryDetails.currencies !== undefined && country.countryDetails.currencies !== null) {
        const currencyKeys = Object.keys(country.countryDetails.currencies);
        setCurrencies(currencyKeys);
      }
    }
  }, [country.detailsStatus, countryName]);

  const currencyParagraph = (currencyKeys) => {
    const currencyDiv = currencyKeys.map((key) => {
        if (country.countryDetails.currencies !== undefined) {
            const currency = country.countryDetails.currencies[key];
            if(currency){
                return <p className="text-white text-right" key={key}>{currency.name} {currency.symbol}</p>;
            }
           return null;
        }
    });
    return currencyDiv;
  };

  return (
    <>
        {country.detailsStatus === "Loading" && <div className='text-white text-xl align-middle h-screen text-center bg-blue-800'>Loading...</div>}
        {country.detailsStatus === "fulfilled" &&
            <div className="bg-blue-700 h-screen">
                <div className="flex flex-col px-3 gap-4 items-center justify-between w-4/5 mx-auto py-12">
                    <img className="w-44 h-auto" src={filteredCountry.flag} alt={`${country.name} flag image`} />
                    <h1 className="text-gray-200 text-right text-shadow text-lato-700 text-2xl">{filteredCountry.officialName}</h1>
                </div>
                <div className="w-full bg-black bg-opacity-20">
                    <p className="ml-4 py-1 text-xs text-lato-400 text-white">COUNTRY INFORMATION - 2023</p>
                </div>
                <section className="w-5/6 m-auto">
                    <div className="flex justify-between my-3">
                        <p className="text-white text-lato-400">Capital</p>
                        <p className="text-white">{country.countryDetails.capital}</p>
                    </div>
                    <div className="flex justify-between my-3 items-center">
                        <p className="text-white text-lato-400 mr-5">Demonym</p>
                        <p className="text-white text-right">{country.countryDetails.demonyms}</p>
                    </div>
                    <div className="flex justify-between my-3">
                        <p className="text-white text-lato-400">Area</p>
                        <p className="text-white">{country.countryDetails.area} Km<sup>2</sup></p>
                    </div>
                    <div className="flex justify-between my-3">
                        <p className="text-white text-lato-400">Time Zones</p>
                        <p className="text-white">{country.countryDetails.timezones}</p>
                    </div>
                    <div className="flex justify-between my-3">
                        <p className="text-white text-lato-400">Map Location</p>
                        <a className="text-blue-200 underline" href={country.countryDetails.mapLocation}>Link to Google Maps</a>
                    </div>
                    <div className="flex justify-between my-3 h-auto items-center">
                        <p className="text-white text-lato-400">Currencies</p>
                        <div className="flex flex-col">
                           {currencyParagraph(currencies)} 
                        </div>
                    </div>
                </section>
                
            </div>
        }
    </>

  )
}

export default DetailsCountry
