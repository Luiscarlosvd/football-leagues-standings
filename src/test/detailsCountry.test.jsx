import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { describe, it, expect } from 'vitest';
import DetailsCountry from '../components/DetailsCountry';
import { BrowserRouter } from 'react-router-dom';

const customMiddleware = [thunk]
const mockStore = configureMockStore(customMiddleware);

describe('DetailsCountry', () => {
  it('should render country details when detailsStatus is fulfilled', () => {
    const store = mockStore({
      countries: {
        detailsStatus: 'fulfilled',
        countryDetails: {
          flag: 'https://restcountries.com/data/usa.svg',
          official: 'United States of America',
          capital: 'Washington, D.C.',
          demonyms: 'American',
          area: 9629091,
          timezones: ['UTC-12:00', 'UTC-11:00'],
          languages: {
            eng: 'English',
            spa: 'Spanish',
          },
          currencies: {
            USD: {
              name: 'United States dollar',
              symbol: '$',
            },
          },
          mapLocation: 'https://www.google.com/maps/place/USA',
        },
      },
    });

    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
            <DetailsCountry />
        </BrowserRouter>
      </Provider>
    );
    
    expect(container).toMatchSnapshot();
  });
});