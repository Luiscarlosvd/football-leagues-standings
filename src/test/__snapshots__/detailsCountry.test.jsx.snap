// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DetailsCountry > should render country details when detailsStatus is fulfilled 1`] = `
<div>
  <div
    class="bg-blue-700 h-full"
  >
    <div
      class="flex flex-col px-3 gap-4 items-center justify-between w-4/5 mx-auto py-12"
    >
      <img
        alt="undefined flag image"
        class="w-44 h-auto mt-10"
        src="https://restcountries.com/data/usa.svg"
      />
      <h1
        class="text-gray-200 text-right text-shadow text-lato-700 text-2xl"
      >
        United States of America
      </h1>
    </div>
    <div
      class="w-full bg-black bg-opacity-20"
    >
      <p
        class="ml-4 py-1 text-xs text-lato-400 text-white"
      >
        COUNTRY INFORMATION - 2023
      </p>
    </div>
    <section
      class="w-full country-information"
    >
      <div
        class="flex justify-between p-5"
      >
        <p
          class="text-white text-lato-700"
        >
          Capital
        </p>
        <p
          class="text-white"
        >
          Washington, D.C.
        </p>
      </div>
      <div
        class="flex justify-between p-5 items-center"
      >
        <p
          class="text-white text-lato-700 mr-5"
        >
          Demonym
        </p>
        <p
          class="text-white text-right"
        >
          American
        </p>
      </div>
      <div
        class="flex justify-between p-5"
      >
        <p
          class="text-white text-lato-700"
        >
          Area
        </p>
        <p
          class="text-white"
        >
          9629091
           Km
          <sup>
            2
          </sup>
        </p>
      </div>
      <div
        class="flex justify-between p-5"
      >
        <p
          class="text-white text-lato-700"
        >
          Time Zones
        </p>
        <p
          class="text-white"
        >
          UTC-12:00
          UTC-11:00
        </p>
      </div>
      <div
        class="flex justify-between p-5 h-auto items-center"
      >
        <p
          class="text-white text-lato-700"
        >
          Language
        </p>
        <div
          class="flex flex-col"
        >
          <p
            class="text-white text-right"
          >
            English
          </p>
          <p
            class="text-white text-right"
          >
            Spanish
          </p>
        </div>
      </div>
      <div
        class="flex justify-between p-5 h-auto items-center"
      >
        <p
          class="text-white text-lato-700"
        >
          Currencies
        </p>
        <div
          class="flex flex-col"
        >
          <p
            class="text-white text-right"
          >
            United States dollar
             
            $
          </p>
        </div>
      </div>
      <div
        class="flex justify-between p-5"
      >
        <p
          class="text-white text-lato-700"
        >
          Map Location
        </p>
        <a
          class="text-blue-200 underline"
          href="https://www.google.com/maps/place/USA"
        >
          Link to Google Maps
        </a>
      </div>
    </section>
  </div>
</div>
`;
